wallet: remove extra wtx lookup in 'AvailableCoins' + several code cleanups.

furszy opened the pull request:
This started in #24845 but grew out of scope of it. 

So, points tackled:

1) Avoid extra `GetWalletTx` lookups inside `AvailableCoins -> IsSpentKey`.
    `IsSpentKey` was receiving the tx hash and index to internally lookup the tx inside the wallet's map. As all the `IsSpentKey` function callers already have the wtx available, them can provide the `scriptPubKey` directly.

2) Most of the time, we call `Wallet::AvailableCoins`, and later on the process, skip the non-spendable coins from the result in subsequent for-loops. So to speedup the process: introduced the ability to filter by "only_spendable" coins inside `Wallet::AvailableCoins` directly.
(the non-spendable coins skip examples are inside `AttemptSelection->GroupOutputs` and `GetAvailableBalance`).

4) Refactored `AvailableCoins` in several ways:

   a) Now it will return a new struct `CoinsResult` instead of receiving the vCoins vector reference (which was being cleared at the beginning of the method anyway). --> this is coming from #24845 but cherry-picked it here too to make the following commits look nicer.

   b) Unified all the 'wtx.tx->vout[I]' calls into a single call (coming from this comment https://github.com/bitcoin/bitcoin/pull/24699#discussion_r854163032). 

5) The wallet `IsLockedCoin` and `IsSpent` methods now accept an `OutPoint` instead of a hash:index. Which let me cleanup a bunch of extra code.

6) Speeded up the wallet 'GetAvailableBalance': filtering `AvailableCoins` by spendable outputs only and using the 'AvailableCoins' retrieved `total_amount` instead of looping over all the retrieved coins once more.

-------------------------------------------------------

Side topic, all this process will look even nicer with #25218

furszy committed:
wallet: return 'CoinsResult' struct in `AvailableCoins`

Instead of accepting a `vCoins` reference that is cleared at the beginning of the method.

Note:
This new struct, down the commits line, will contain other `AvailableCoins` useful results.

furszy committed:
wallet: AvailableCoins, don't call 'wtx.tx->vout[i]' multiple times

furszy committed:
wallet: IsLockedCoin, 'COutPoint' arg instead of (hash, index)

furszy committed:
wallet: IsSpent, 'COutPoint' arg instead of (hash, index)

furszy committed:
wallet: decouple IsSpentKey(scriptPubKey) from IsSpentKey(hash, n)

This will be used in a follow-up commit to prevent extra 'GetWalletTx' lookups if the function caller already have the wtx and can just provide the scriptPubKey directly.

furszy committed:
wallet: avoid extra IsSpentKey -> GetWalletTx lookups

furszy committed:
wallet: remove unused IsSpentKey(hash, index) method

furszy committed:
wallet: GetAvailableBalance, remove double walk-through every available coin

Filtering `AvailableCoins` by spendable outputs only and using the retrieved total_amount.

jonatack commented:
@furszy It looks like the qt wallet test is failing in the CI. You can run `./src/qt/test/test_bitcoin-qt` in your dev environment to reproduce and debug the issue.

furszy commented:
Ah, thanks @jonatack 👌. Just noticed that the GUI `WalletTests` isn't run by default in my OS under `make check`.

It's an easy fix; as the test doesn't start the balance polling timer, the model does not have the balance cached so.. as 5babfb7 make uses of it, the creation process fails with a not-enough balance error.

Will push the fix and expand + improve the test coverage a bit more.
And.. most likely will decouple that last commit into a separate PR in few days. Have some further improvements locally for the wallet model area that can package out all together that aren't really part of the scope of this PR.

furszy commented:
> 6) The GUI `WalletModel::prepareTransaction` will now use the cached balance to make the funds availability check if the user didn't manually selected outputs to spend. This will avoid an entire on-demand recalculation of the available balance which is quite expensive as the wallet has to walk through the entire txs map.

Dropped the last two commits that were implementing the point (6) that is above. They are now part of https://github.com/bitcoin-core/gui/pull/598

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -225,17 +223,13 @@ CoinsResult AvailableCoins(const CWallet& wallet,
-    }
-    return balance;
+    // Copy to filter by spendable coins only
+    CCoinControl coinControl = _coinControl ? *_coinControl : CCoinControl();
```
In 2cdd2a1a5498cb8767799dd26a4082e9d3b064ee "wallet: speedup 'GetAvailableBalance' filtering by spendable outputs only and using the 'AvailableCoins' retrieved total_amount."

What is the purpose of the rename and making an empty `CCoinControl` here? `AvailableCoins` can take `nullptr` for `coinControl`.

furszy replied:
The previous flow was:
```c++
std::vector<COutput> vCoins;
AvailableCoins(wallet, vCoins, coinControl);
for (const COutput& out : vCoins) {
     if (out.spendable) { --> // this is the reason
         balance += out.txout.nValue;
     }
}
```

To remove the second unnecessary for-loop, as the function arg is `const CCoinControl*`, in order to continue having the same calculation (skipping the non-spendable coins), needed to create a local copy of the `coinControl` arg and set the new flag `m_include_only_spendable_outputs` to true (which by default is false).

So the coins vector and the total balance returned by `AvailableCoins ` has all, and only, the spendable coins directly.

achow101 reviewed (commented):
I don't quite understand the purpose of the first commit. It doesn't seem like it materially improves anything?

furszy commented:
Hey achow101, thanks for the review!

> I don't quite understand the purpose of the first commit. It doesn't seem like it materially improves anything?

We are skipping the non-spendable coins that appear in `vCoins` (`AvailableCoins` result) later, in several parts of the `CreateTransaction` and `GetBalance` flows:

1.  `GetAvailableBalance` was (1) getting all the available coins calling `AvailableCoins` and, right away, walking through the entire vector, skipping the non-spendable coins, to calculate the total balance. 

2. Inside `CreateTransactionInternal` —> `SelectCoins(vCoins,...)`, we have several calls to `AttemptSelection` which, on each of them internally, we call twice to `GroupOutputs` which internally has two for-loops over the entire `vCoins` vector that skip the non-spendable coins.

So, the purpose is not add the non-spendable coins into the `AvailableCoins` result (`vCoins`) in the first place for the processes that aren’t using them at all, so we don’t waste resources skipping them later so many times.

Note: this speedup is for all the processes that calls to `CreateTransaction` and `GetBalance*` internally.

furszy commented:
Rebased after #25083 merge. Conflicts solved.

furszy committed:
wallet: add 'only_spendable' filter to AvailableCoins

We are skipping the non-spendable coins that appear in vCoins ('AvailableCoins' result) later, in several parts of the CreateTransaction and GetBalance flows:

GetAvailableBalance (1) gets all the available coins calling AvailableCoins and, right away, walk through the entire vector, skipping the non-spendable coins, to calculate the total balance.

Inside CreateTransactionInternal —> SelectCoins(vCoins,...), we have several calls to AttemptSelection which, on each of them internally, we call twice to GroupOutputs which internally has two for-loops over the entire vCoins vector that skip the non-spendable coins.

So, Purpose is not add the non-spendable coins into the AvailableCoins result (vCoins) in the first place for the processes that aren’t using them at all, so we don’t waste resources skipping them later so many times.

Note: this speedup is for all the processes that call to CreateTransaction and GetBalance* internally.

furszy commented:
Ready, feedback applied:

- Moved `only_spendable` from be a `CoinControl` field to be a `AvailableCoins` param.
- Nits tackled.

theStack commented:
Concept ACK

theStack reviewed (commented):
LGTM, changes all make sense to me, also thanks to chat we had in IRC :)

Left a few nits below. Also, can you change the commit texts to have a shorter line length? This makes the git history more readable. Breaking at 72-80 characters seem to be good values.

furszy commented on src/wallet/spend.cpp:
```diff
@@ -159,76 +165,82 @@ void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const C
-        }
-    }
-    return balance;
+    return AvailableCoins(wallet,
+                          coinControl,
+                          std::nullopt, /*feerate=*/
+                          1,            /*nMinimumAmount*/
+                          MAX_MONEY,    /*nMaximumAmount*/
+                          MAX_MONEY,    /*nMinimumSumAmount*/
+                          0,            /*nMaximumCount*/
+                          true          /*spendable_only*/
+                          ).total_amount;
```
You found an existing issue in the RPC `sendall` command :).

When `send_max` is set, `sendall` uses all the coins returned from `AvailableCoins` as inputs for a new transaction, without checking if the coin is spendable/solvable first.

So.. this command should be always failing in master if the user sets the `send_max` param and the wallet has at least one non-spendable or non-solvable coin.

Note:
This is easily solved inputting the new `spendable_only=false` flag to `AvailableCoins` here.

furszy commented:
thanks for the feedback theStack!

Applied the following changes:

1) Re-wrote commits titles to be 72-80 characters long.
2) Moved the `only_spendable` arg of `AvailableCoins` to be true by default.
3) By having `spendable_only=true` by default, fixed an existent hidden issue in `sendall` RPC command -> [description](https://github.com/bitcoin/bitcoin/pull/25005#discussion_r892438732)

brunoerg commented:
Concept ACK

achow101 commented:
ACK fd5c996d1609e6f88769f6f3ef0c322e3435b3aa

brunoerg reviewed (approved):
crACK fd5c996d1609e6f88769f6f3ef0c322e3435b3aa

w0xlt reviewed (approved):
Code Review ACK https://github.com/bitcoin/bitcoin/pull/25005/commits/fd5c996d1609e6f88769f6f3ef0c322e3435b3aa

Clear improvement in code readability and `Available Coins()` logic.

stickies-v reviewed (commented):
Post-merge approach ACK fd5c996, nice refactor!

I did have a couple of nits/style suggestions during review, none of them crucial. I'll post them anyway since you may want to consider them in future PRs.

One thing I wonder (just out of curiosity) is why for `IsSpent()` and `IsLockedCoin()` you refactored the function signature in a single commit, whereas for `IsSpentKey()` you did it in three commits? I think both are done elegantly, just curious if there's a reason behind the (apparent) inconsistency.

MarcoFalke commented:
https://cirrus-ci.com/task/4885942589718528?logs=ci#L6616

```
wallet/spend.cpp:237:41: error: argument name 'feerate' in comment does not match parameter name 'nMinimumAmount' [bugprone-argument-comment,-warnings-as-errors]
                          std::nullopt, /*feerate=*/
                                        ^
./wallet/spend.h:49:43: note: 'nMinimumAmount' declared here
                           const CAmount& nMinimumAmount = 1,
                                          ^
wallet/spend.cpp:87:13: note: actual callee ('AvailableCoins') is declared here
CoinsResult AvailableCoins(const CWallet& wallet,
            ^
